### PR TITLE
Implement leaf first traversal

### PR DESCRIPTION
**Pull request by DavidDudson**
Needed for scalagen

Will not affect existing code.

Note: Implicit param has been added to the traverse method, so this is a bin-compat issue.

### CODE REVIEW

**Comment by DavidDudson** [comment]
This fixes #1191

File: scalameta/transversers/shared/src/test/scala/meta/transversers/OrderTests.scala [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+      "scala.meta.Template$TemplateImpl",
+      "scala.meta.Defn$Class$DefnClassImpl")
+
+  val input = q"class Foo { def bar = 1 }"
+
+  test("Traverse visits roots first") {
+    val visited: mutable.ArrayBuffer[String] = mutable.ArrayBuffer()
+
+    input.traverse {
+      case a => visited.append(a.getClass.getName)
```
Comment (olafurpg) [inline scalameta/transversers/shared/src/test/scala/meta/transversers/OrderTests.scala]: FYI `.productPrefix` will give you a cleaner name per tree node

Comment (DavidDudson) [reply scalameta/transversers/shared/src/test/scala/meta/transversers/OrderTests.scala]: Nice! Usually I use getClass.getSimpleName but of course that fails here.

**Comment by DavidDudson** [comment]
Let's leave this PR outstanding for now. I'll try and whip something up for `scalagen` specfically for now.

If possible, I would like to convert the AST/Traversal macros to generators before scalameta 3.0. That way the generators would be a lot more approachable than the current macro based code for modifying.

File: scalameta/transversers/shared/src/main/scala/scala/meta/transversers/Api.scala [hunk 2]
```diff
@@ -2,33 +2,58 @@ package scala.meta
 package transversers
 
 private[meta] trait Api {
+
+  sealed trait TraversalOrder
+  case object LeafFirst extends TraversalOrder
+  case object RootFirst extends TraversalOrder
```
Comment (xeno-by) [inline scalameta/transversers/shared/src/main/scala/scala/meta/transversers/Api.scala]: How about factoring out these definitions into `TraversalOrder.scala` and then aliasing them here if you feel that these guys belong to the public API? This is the pattern that we're using in other modules of Scalameta.

File: scalameta/transversers/shared/src/test/scala/meta/transversers/OrderTests.scala [hunk 3]
```diff
@@ -0,0 +1,109 @@
+package scala.meta.transversers
```
Comment (xeno-by) [inline scalameta/transversers/shared/src/test/scala/meta/transversers/OrderTests.scala]: Recently we have moved the overwhelming majority of tests into a single big project at `scalameta/tests`. Let's continue doing this in future.

**Comment by olafurpg** [comment]
@DavidDudson if we're gonna whip up something in scalagen I propose we close this PR. I'd love to expose more configurable traversal strategies in the main scalameta API but I think we should 
- consider how to do it without breaking the existing API
- get more hands-on experience with it in scalagen
In scalafix, I also have a lot of scalameta utilities that should probably go upstream at some point.

**Comment by DavidDudson** [comment]
Fair enough

**Pull request closed without merging**